File: 08/VMparser.java
import java.io.File;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.file.Files;
import java.util.ArrayList;
import java.util.List;

/**
 * A class that manages parsing of a .vm file.
 */
public class VMparser {
	private static final String[] ARITHMIC_SIGNS = new String[] {"add", "sub", "neg", "eq", "gt",
															     "lt", "and", "or", "not"};
	private static final String POP = "pop";
	private static final String PUSH = "push";
	private static final String LABEL = "label";
	private static final String GOTO = "goto";
	private static final String IF_GOTO = "if-goto";
	private static final String FUNCTION = "function";
	private static final String CALL = "call";
	private static final String RETURN = "return";
	private static final String UTF8 = "UTF-8";
	private static final String COMMENT_PREFIX = "//";
	private static final String WHITESPACE = "\\s";
	private static final char DOT = '.';

	public enum CommandType {C_ARITHMETIC, C_PUSH, C_POP, C_LABEL, C_GOTO, C_IF_GOTO, C_FUNCTION,
							 C_RETURN, C_CALL, FAILURE}

	private int currIdx = 0;
	private File myFile = null;
	private ArrayList<String> myLines = new ArrayList<>();

	/**
	 * Class Constructor.
	 * @param file A file object that encapsulate a .vm file.
	 */
	public VMparser(File file) {
		this.myFile = file;
	}

	/**
	 * Returns the .vm file's name.
	 * @return String which is this file's name.
	 */
	public String getFileName() {
		return this.myFile.getName().substring(0, myFile.getName().lastIndexOf(DOT));
	}

	/**
	 * Method which returns a boolean which is an indicator whether this parser has finished parsing
	 * the .vm file or not.
	 * @return true iff the parser finished parsing the file.
	 */
	public boolean hasFinished() {
		return this.currIdx > this.myLines.size()-1;
	}

	/**
	 * Method that returns the current command the parser holds.
	 * @return String which is the current line/command this parser holds.
	 */
	public String getCurrentCommand() {
		return myLines.get(currIdx);
	}

	/**
	 * Method that advances this parser's current line index.
	 */
	public void advance() {
		++this.currIdx;
	}

	/**
	 * Method that removes all comment or blank lines, as well as comments in lines who have
	 * comments at their ends.
	 * @throws IOException
	 */
	public void removedExcessLines() throws IOException {
		Charset charSet = Charset.forName(UTF8);
		List<String> lines = Files.readAllLines(myFile.toPath(), charSet);
		for (String line : lines) {
			String[] noCommentsLine = line.split(COMMENT_PREFIX);
			String noSpacesLine = noCommentsLine[0].trim();
			if (noSpacesLine.isEmpty()) continue;
			this.myLines.add(noSpacesLine);
		}
	}

	/**
	 * Method that returns the current command-line's type, which can be one of the types in the
	 * CommandType enum class.
	 * @return CommandType an enum class that holds possible command types.
	 */
	public CommandType getCmdType() {
		for (String arithmicSign : ARITHMIC_SIGNS) {
			if (myLines.get(currIdx).contains(arithmicSign)) return CommandType.C_ARITHMETIC;
		}
		if (myLines.get(currIdx).contains(POP)) return CommandType.C_POP;
		else if (myLines.get(currIdx).contains(PUSH)) return CommandType.C_PUSH;
		else if (myLines.get(currIdx).contains(LABEL)) return CommandType.C_LABEL;
		else if (myLines.get(currIdx).contains(GOTO)) return CommandType.C_GOTO;
		else if (myLines.get(currIdx).contains(IF_GOTO)) return CommandType.C_IF_GOTO;
		else if (myLines.get(currIdx).contains(FUNCTION)) return CommandType.C_FUNCTION;
		else if (myLines.get(currIdx).contains(CALL)) return CommandType.C_CALL;
		else if (myLines.get(currIdx).contains(RETURN)) return CommandType.C_RETURN;
		else {
			return CommandType.FAILURE;
		}
	}

	/**
	 * Method that returns the first argument of non-arithmetic command-lines.
	 * @return String that represent the first argument of a non-arithmetic command-line.
	 */
	public String getFirstArg() {
		String[] lineWords = myLines.get(currIdx).split(WHITESPACE);
		CommandType currCmd = this.getCmdType();
		if (currCmd == CommandType.C_ARITHMETIC) return lineWords[0];
		return lineWords[1];
	}

	/**
	 * Method that returns the second argument of non-arithmetic command-lines.
	 * @return String that represent the second argument of a non-arithmetic command-line.
	 */
	public String getSecondArg() {
		String[] lineWords = myLines.get(currIdx).split(WHITESPACE);
		return lineWords[2];
	}

}

File: 08/FunctionCalls/FibonacciElement/VMparser.py
__author__ = 'Or Keren'


class Command():
    """
    An enum class that represents tokens for each type of vm command
    """
    C_NULL = 0
    C_ARITHMETIC = 1
    C_PUSH = 2
    C_POP = 3
    C_LABEL = 4
    C_GOTO = 5
    C_IF = 6
    C_FUNCTION = 7
    C_RETURN = 8
    C_CALL = 9


class Parser:
    """
    The Parser class, the class holds all the lines of code that are in the VM file. It allows the translator to
    traverse the file line-by-line
    """
    def __init__(self, file_name):
        """
        The constructor of the class, initializes the internal parameters
        :param file_name: The file name the parser is attached to
        :return: N/A
        """
        file = open(file_name, 'r')
        code = file.readlines()
        file.close()

        self._code = []
        self.remove_excess(code)
        self._num_of_commands = len(self._code)
        self._counter = 0
        self._command = None
        self._file_name = file_name

    def remove_excess(self, code):
        """
        Removes excess comments from all the code
        :param code: All the lines of code in the file
        :return: N/A
        """
        for line in code:
            line.replace('\r', '')
            line = ' '.join(line.rsplit())

            if '//' in line:
                line = line[:line.find('//')]
            if line != '':
                self._code.append(line)

    def can_continue(self):
        """
        Returns if there are more lines that the parser can traverse over
        :return: true if the number of total lines in the VM file is bigger than the current line counter
        """
        # Returns whether we still have commands that need to be read
        return self._num_of_commands > self._counter

    def advance(self):
        """
        Advances the parser by one line
        :return: N/A
        """
        # Advances the "pointer" that points to the current command we are reading in the file
        self._command = self._code[self._counter]
        self._counter += 1

    def get_command_type(self):
        """
        Tokenizes the current command line into the relevant enum
        :return: The enum representing the command of the current VM line
        """
        return {
        "add": Command.C_ARITHMETIC,
        "sub": Command.C_ARITHMETIC,
        "neg": Command.C_ARITHMETIC,
        "eq": Command.C_ARITHMETIC,
        "gt": Command.C_ARITHMETIC,
        "lt": Command.C_ARITHMETIC,
        "and": Command.C_ARITHMETIC,
        "or": Command.C_ARITHMETIC,
        "not": Command.C_ARITHMETIC,
        "pop": Command.C_POP,
        "push": Command.C_PUSH,
        "label": Command.C_LABEL,
        "goto": Command.C_GOTO,
        "if-goto": Command.C_IF,
        "function": Command.C_FUNCTION,
        "call": Command.C_CALL,
        "return": Command.C_RETURN
    }.get(self._command.rsplit()[0], Command.C_NULL)

    def get_file_name(self):
        """
        :return: Returns the file name the parser is attached to
        """
        return self._file_name

    def get_curr_command(self):
        """
        :return: Returns the current command (i.e the line of code in the vm file)
        """
        return self._command

    def get_first_arg(self):
        """
        :return: Returns the first argument of the VM command
        """
        split_line = self._command.rsplit()
        command_type = self.get_command_type()

        if command_type in [Command.C_ARITHMETIC,Command.C_RETURN]:
            return split_line[0]
        else:
            return split_line[1]

    def get_second_arg(self):
        """
        :return: Returns the second argument of the VM command
        """
        split_line = self._command.rsplit()

        # If there is a second command it is always an integer value
        return int(split_line[2] if len(split_line)>=3 else -1)